### PR TITLE
Wip/hughsie/hsi docbook

### PR DESCRIPTION
**Pull request by hughsie**

### CODE REVIEW

**Comment by hughsie** [comment]
Screenshot from 2020-07-27 19-42-10
Screenshot from 2020-07-27 19-42-27

File: docs/hsi.xml [hunk 1]
```diff
@@ -0,0 +1,875 @@
...
+          <firstname>Alex</firstname>
+          <surname>Matrosov</surname>
+        </personname>
+        <affiliation>
+          <orgname>Nvidia</orgname>
+        </affiliation>
+      </author>
+    </authorgroup>
+  </info>
+-->
```
Comment (superm1) [inline docs/hsi.xml:55]: If these weren't commented out, how do they actually render?  Or do they?

File: docs/hsi.xml [hunk 2]
```diff
@@ -0,0 +1,875 @@
...
+        from signed reports uploaded to the LVFS and from analyzing firmware binaries.
+        However this information alone does not explain firmware security to the
+        user in a way they can actually interpret.
+      </para>
+    </refsect1>
+
+    <refsect2 id="existing-tooling">
+      <title>Existing Tooling</title>
+      <para>
+        Today figuring out the true security of your hardware and firmware
```
Comment (superm1) [inline docs/hsi.xml:85]: The word "Today" doesn't generally age well in something like this.

File: docs/hsi.xml [hunk 3]
```diff
@@ -0,0 +1,875 @@
...
+        people buying hardware.
+        The service already knows a huge amount of information about machines
+        from signed reports uploaded to the LVFS and from analyzing firmware binaries.
+        However this information alone does not explain firmware security to the
+        user in a way they can actually interpret.
+      </para>
+    </refsect1>
+
+    <refsect2 id="existing-tooling">
+      <title>Existing Tooling</title>
```
Comment (superm1) [inline docs/hsi.xml:83]: Right now existing means "pre-fwupd 1.5.x tooling".

Maybe this section needs to be renamed to "Other tools"?

File: docs/hsi.xml [hunk 4]
```diff
@@ -0,0 +1,875 @@
...
+    </refsect1>
+
+    <refsect2 id="runtime-behaviour">
+      <title>Runtime Behavior</title>
+      <para>
+        Orthogonal to the security features provided by the firmware there are
+        other security considerations related to the firmware which may require
+        internet access to discover or that runtime OS changes directly affect
+        the security of the firmware.
+        It would not make sense to have <emphasis>“have updates on the LVFS”</emphasis>
```
Comment (superm1) [inline docs/hsi.xml:140]: with the docbook markup, does it still need `"`

File: docs/hsi.xml [hunk 5]
```diff
@@ -0,0 +1,875 @@
...
+        as a requirement for a specific security level as this would mean
+        offline the platform might be a higher level initially but as soon as
+        it is brought online it is downgraded which would be really confusing to
+        users.
+        The “core” security level will not change at runtime, but the suffix may.
```
Comment (superm1) [inline docs/hsi.xml:145]: Operating System runtime

File: docs/hsi.xml [hunk 6]
```diff
@@ -0,0 +1,875 @@
...
+      </para>
+    </refsect2>
+
+    <refsect2 id="hsi-level0">
+      <title>HSI:0 (Insecure)</title>
+      <para>
+        The lowest security level with very few detected firmware protections.
```
Comment (superm1) [inline docs/hsi.xml:152]: I would suggest `s/very few/little or no/`

File: docs/hsi.xml [hunk 7]
```diff
@@ -0,0 +1,875 @@
...
+      <para>
+        Updates available on the <ulink url="https://fwupd.org/">
+        Linux Vendor Firmware Service </ulink> which are less than 12 months old.
+      </para>
+    </refsect3>
+
+    <refsect3 id="runtime-a">
+      <title>HSI Runtime Suffix <code>A</code></title>
+      <para>
+        Attestation data is available to verify the current PCR0 TPM hash.
```
Comment (superm1) [inline docs/hsi.xml:207]: Something about this being UFEI specific and linking to an explanation of PCR0 would be useful here.

File: docs/hsi.xml [hunk 8]
```diff
@@ -0,0 +1,875 @@
...
+        <listitem>
+          <para>
+            The kernel is not not <ulink url="https://mjg59.dreamwidth.org/50577.html">
+            locked down</ulink>. <emphasis>v1.5.0</emphasis>
+          </para>
+        </listitem>
+        <listitem>
+          <para>
+            Unencrypted <ulink url="https://wiki.archlinux.org/index.php/Dm-crypt/Swap_encryption">
+            swap partition</ulink>. <emphasis>v1.5.0</emphasis>
```
Comment (superm1) [inline docs/hsi.xml:238]: Maybe worth including a link to your blog for this too?

File: docs/hsi.xml [hunk 9]
```diff
@@ -0,0 +1,875 @@
...
+            </listitem>
+          </itemizedlist>
+        </para>
+      </note>
+    </refsect3>
+
+    <refsect3 id="org.fwupd.hsi.Spi.Ble">
+      <title>BIOS Lock Enable (BLE)</title>
+      <para>
+        If the lock bit is set then SMIs are raised when setting BIOS Write Enable.
```
Comment (superm1) [inline docs/hsi.xml:289]: Probably need to expand SMI as this is the first use of it.

File: docs/hsi.xml (hunk 7 above)
Comment (superm1) [reply docs/hsi.xml:207]: Also this is the first time a TPM is mentioned and it might need to be explained too.

File: docs/hsi.xml [hunk 10]
```diff
@@ -0,0 +1,875 @@
...
+          </itemizedlist>
+        </para>
+      </note>
+    </refsect3>
+
+    <refsect3 id="org.fwupd.hsi.Mei.Version">
+      <title>CSME Version</title>
+      <para>
+        Converged Security and Manageability Engine is a standalone management
+        module that can manage and control all the local devices without the host
```
Comment (superm1) [inline docs/hsi.xml:468]: That's a bit of an overreach.  It only controls very specific devices.

File: docs/hsi.xml [hunk 11]
```diff
@@ -0,0 +1,875 @@
...
+              <ulink url="https://www.kernel.org/doc/html/latest/security/tpm/tpm_event_log.html">
+                Linux Kernel TPM Documentation
+              </ulink>
+            </listitem>
+          </itemizedlist>
+        </para>
+      </note>
+    </refsect3>
+    <refsect3 id="org.fwupd.hsi.AcpiDmar">
+      <title>Kernel DMA protection</title>
```
Comment (superm1) [inline docs/hsi.xml:570]: This is pre-boot specifically, no?

File: docs/hsi.xml [hunk 12]
```diff
@@ -0,0 +1,875 @@
...
+      </para>
+    </refsect1>
+
+    <refsect2 id="ommissions">
+      <title>Intentional Omissions</title>
+    </refsect2>
+    <refsect3 id="intel-txt">
+      <title>Intel TXT</title>
+      <para>
+        This security functionality requires a binary blob to operate.
```
Comment (superm1) [inline docs/hsi.xml:839]: eh?  https://github.com/TrenchBoot

File: docs/hsi.xml [hunk 13]
```diff
@@ -0,0 +1,875 @@
...
+      <title>Intel SGX</title>
+      <para>
+        This is not widely used as it has several high severity security issues.
+      </para>
+    </refsect3>
+    <refsect3 id="intel-mpx">
+      <title>Intel MPX</title>
+      <para>
+        MPX support has now been removed from popular compilers and the kernel
+        several years ago and it is now considered obsolete.
```
Comment (superm1) [inline docs/hsi.xml:853]: this statement won't age well - probably need a specific time frame to include

File: docs/hsi.xml (hunk 1 above)
Comment (hughsie) [reply docs/hsi.xml:55]: I commented them out for two reasons:

 * I'd like people to sign-off the final version, possibly when the ARM/AMD stuff has landed too
 * The rendering of the names+vendors is simply rubbish in the gtk-doc style sheet -- if we **do** want the author block then I might resort to `<para>` and `<itemizedlist>` as a hack.

Screenshot_2020-07-28 Host Security ID Specification fwupd Reference Manual

File: docs/hsi.xml (hunk 2 above)
Comment (hughsie) [reply docs/hsi.xml:85]: how about,

        Traditionally, figuring out the true security of your hardware and firmware
        requires sifting through the marketing documentation provided by the
        OEM and in many cases just “trusting” they did it right.

File: docs/hsi.xml (hunk 3 above)
Comment (hughsie) [reply docs/hsi.xml:83]: Agree, fixed.

File: docs/hsi.xml (hunk 4 above)
Comment (hughsie) [reply docs/hsi.xml:140]: Good point, fixed.

File: docs/hsi.xml (hunk 8 above)
Comment (hughsie) [reply docs/hsi.xml:238]: I considered it, but I wanted an independent source really. Like a wikipedia page, if it existed.

File: docs/hsi.xml (hunk 10 above)
Comment (hughsie) [reply docs/hsi.xml:468]: How about ... `manage and control some local devices`

File: docs/hsi.xml (hunk 13 above)
Comment (hughsie) [reply docs/hsi.xml:853]: How about

        MPX support was removed from GCC and the Linux kernel in 2019 and it is
        now considered obsolete.

File: docs/hsi.xml (hunk 9 above)
Comment (hughsie) [reply docs/hsi.xml:289]: like?

        If the lock bit is set then System Management Interrupts (SMIs) are
        raised when setting BIOS Write Enable.

File: docs/hsi.xml (hunk 7 above)
Comment (hughsie) [reply docs/hsi.xml:207]: How about

        Attestation data is available to verify the current system firmware.
        For most UEFI platforms, this is usually the Trusted Platform Module (TPM)
        PCR0 hash, but other attestation checksums could be used.

File: docs/hsi.xml (hunk 11 above)
Comment (hughsie) [reply docs/hsi.xml:570]: I think this is a good example of where the spec and fwupd were disagreeing on what stuff actually meant; and one of the big reasons I wanted to bring it all into one place (the other, being able to add URLs to the various HSI attrs to link back to the spec...).

This is of course pre-boot IOMMU. How about this:

      <title>Pre-boot DMA protection</title>
      <para>
        The IOMMU on modern systems is used to mitigate against DMA attacks.
        All I/O for devices capable of DMA is mapped into a private virtual
        memory region.
        The ACPI DMAR table is used to set up pre-boot DMA protection which
        eliminates some firmware attacks.
      </para>

File: docs/hsi.xml (hunk 12 above)
Comment (hughsie) [reply docs/hsi.xml:839]: I'll admit I didn't know how TrenchBoot and TXT operate together. Shall I just remove this TXT section completely for now?

File: docs/hsi.xml (hunk 1 above)
Comment (superm1) [reply docs/hsi.xml:55]: Ah OK, thanks.

File: docs/hsi.xml (hunk 2 above)
Comment (superm1) [reply docs/hsi.xml:85]: Yeah, that's better

File: docs/hsi.xml (hunk 7 above)
Comment (superm1) [reply docs/hsi.xml:207]: Sounds good, and it's good to disambiguate that something non-UEFI may exist some day.

File: docs/hsi.xml (hunk 8 above)
Comment (superm1) [reply docs/hsi.xml:238]: OK.

File: docs/hsi.xml (hunk 9 above)
Comment (superm1) [reply docs/hsi.xml:289]: 👍

File: docs/hsi.xml (hunk 10 above)
Comment (superm1) [reply docs/hsi.xml:468]: Yes

File: docs/hsi.xml (hunk 11 above)
Comment (superm1) [reply docs/hsi.xml:570]: 👍

File: docs/hsi.xml (hunk 12 above)
Comment (superm1) [reply docs/hsi.xml:839]: I think that's better for now, yes.  But also it's worth looking into, it's how TXT can come to the masses.

File: docs/hsi.xml (hunk 13 above)
Comment (superm1) [reply docs/hsi.xml:853]: 👍

**Comment by hughsie** [comment]
@superm1 should I push this to master now? I'd like to propose some new tests in a few days and I'd like to use the URL thing to point to the spec so I can add the spec entry at the same time.

**Pull request merged**